subxt/src/metadata/metadata_type.rs

```diff
@@ -120,12 +144,90 @@ impl Metadata {
 
     /// Resolve a type definition.
     pub fn resolve_type(&self, id: u32) -> Option<&Type<PortableForm>> {
-        self.metadata.types.resolve(id)
+        self.inner.metadata.types.resolve(id)
     }
 
     /// Return the runtime metadata.
-    pub fn runtime_metadata(&self) -> &RuntimeMetadataLastVersion {
-        &self.metadata
+    pub fn runtime_metadata(&self) -> &RuntimeMetadataV14 {
+        &self.inner.metadata
+    }
+
+    /// Obtain the unique hash for a specific storage entry.
+    pub fn storage_hash<S: crate::StorageEntry>(
+        &self,
+    ) -> Result<[u8; 32], MetadataError> {
+        self.inner
+            .cached_storage_hashes
+            .get_or_insert(S::PALLET, S::STORAGE, || {
+                subxt_metadata::get_storage_hash(
+                    &self.inner.metadata,
+                    S::PALLET,
+                    S::STORAGE,
+                )
+                .map_err(|e| {
+                    match e {
+                        subxt_metadata::NotFound::Pallet => MetadataError::PalletNotFound,
+                        subxt_metadata::NotFound::Item => MetadataError::StorageNotFound,
+                    }
+                })
+            })
+    }
+
+    /// Obtain the unique hash for a constant.
+    pub fn constant_hash(
+        &self,
+        pallet: &str,
+        constant: &str,
+    ) -> Result<[u8; 32], MetadataError> {
+        self.inner
+            .cached_constant_hashes
+            .get_or_insert(pallet, constant, || {
+                subxt_metadata::get_constant_hash(&self.inner.metadata, pallet, constant)
+                    .map_err(|e| {
+                        match e {
+                            subxt_metadata::NotFound::Pallet => {
+                                MetadataError::PalletNotFound
+                            }
+                            subxt_metadata::NotFound::Item => {
+                                MetadataError::ConstantNotFound
+                            }
+                        }
+                    })
+            })
+    }
+
+    /// Obtain the unique hash for a call.
+    pub fn call_hash<C: crate::Call>(&self) -> Result<[u8; 32], MetadataError> {
+        self.inner
+            .cached_call_hashes
+            .get_or_insert(C::PALLET, C::FUNCTION, || {
+                subxt_metadata::get_call_hash(
+                    &self.inner.metadata,
+                    C::PALLET,
+                    C::FUNCTION,
+                )
+                .map_err(|e| {
+                    match e {
+                        subxt_metadata::NotFound::Pallet => MetadataError::PalletNotFound,
```

Comment:
maybe a `From` impl could get rid of this repetitive matching?

Reply:
It's slightly different from the above one though; PalletNotFound and CallNotFound here vs PalletNotFound and ConstantNotFound above :)

Reply:
fair enough, sloppy of me not to catch that :)